docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,315 @@
+    Cinder and NFS volumes
+
+In order to do this, the identity of the provisioner must be a property of the
+storage class (as opposed to implicit, as it is in the alpha feature).
+
+#### Parameterizing provisioners
+
+It will be common for a cluster to have multiple storage classes which
+leverage the same provisioner to create volumes with different characteristics
+for the different classes.  For example, an administrator running a cluster
+hosted in AWS might want to have two storage classes, `slow` and `fast`, that
+both leverage the same EBS provisioner to provision spinning disk volumes and
+provisioned-iops volumes, respectively.  From this, it follows that:
+
+1.  Storage classes should be able to hold a set of provisioner parameters
+2.  The provisioner should be passed the claim and the details of the storage
+    class it is provisioning for
+
+#### Out-of-tree provisioners
+
+One of our goals is to enable administrators to create out-of-tree
+provisioners, that is, provisioners whose code does not live in the Kubernetes
+project.  Our experience since the 1.2 release with dynamic provisioning has
+shown that it is impossible to anticipate every aspect and manner of
+provisioning that administrators will want to perform.  The proposed design
+should not prevent future work to allow out-of-tree provisioners.
+
+## Analysis
+
+### Summary of alpha functionality
+
+The alpha functionality for dynamic provisioning works as follows:
+
+1.  A cluster has only a single provisioner at a time, dictated by the cloud
+    provider
+2.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.  If a suitable volume is found, the
+    controller binds the claim and volume together and carries on normally.
+3.  If no volume is found for the claim, and the claim has the annotation
+    `volume.alpha.kubernetes.io/storage-class`, the provisioner is invoked
+    to provision a volume to fill that claim
+4.  All provisioners are in-tree; they implement an interface called
+    `ProvisionableVolumePlugin` which has a method, `NewProvisioner()`,
+    that returns a new `Provisioner` instance
+5.  The controller calls the provisioner's `Provision()` method; `Provision`
+    is responsible for provisioning the volume in the cloud provider and
+    returns an `api.PersistentVolume` and an error
+6.  If `Provision` returns an error, the controller stops trying to provision
+7.  If `Provision` returns no error, the controller creates the returned
+    `api.PersistentVolume`, already bound to the claim
+  1.  If the create operation fails, it is retried
+  2.  If the create operation never succeeds, the controller attempts to delete
+      the provisioned volume and creates an event on the claim
+
+### When should we provision volumes?
+
+It is sensible to provision volumes only when an existing volume cannot
+satisfy a claim.  This ensures that existing, already allocated resources are
```

Comment:
We want to be careful to not over-design the selection semantics.  It's
really easy to end up with arbitrary boolean expressions, which we belive
doesn't provide enough value to justify the complexity.  There is a very
real issue in the way we define matches today, though, which is surprising
to some people, and can be very inefficient.

On Fri, Jul 8, 2016 at 8:56 AM, Jing Xu <notifications@github.com> wrote:

> In docs/proposals/volume-provisioning.md
> <https://github.com/kubernetes/kubernetes/pull/26908#discussion_r70096996>
> :
>
> > +    `ProvisionableVolumePlugin` which has a method, `NewProvisioner()`,
> > +    that returns a new `Provisioner` instance
> > +5.  The controller calls the provisioner's `Provision()` method; `Provision`
> > +    is responsible for provisioning the volume in the cloud provider and
> > +    returns an `api.PersistentVolume` and an error
> > +6.  If `Provision` returns an error, the controller stops trying to provision
> > +7.  If `Provision` returns no error, the controller creates the returned
> > +    `api.PersistentVolume`, already bound to the claim
> > +  1.  If the create operation fails, it is retried
> > +  2.  If the create operation never succeeds, the controller attempts to delete
> > +      the provisioned volume and creates an event on the claim
> > +
> > +### When should we provision volumes?
> > +
> > +It is sensible to provision volumes only when an existing volume cannot
> > +satisfy a claim.  This ensures that existing, already allocated resources are
>
> Is that possible to allow user to explicitly specify some operators like
> larger than, less than, or inside of a range. For capacity requirement, it
> would be clear what is satisfied. Also NOT operator might be useful too in
> selecting storage class, I think.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/26908/files/9b8346a996ac2a02773987370506b76043c8c44b#r70096996>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFVgVExwY22WOCyW8UdIwCCDJFCQgTWbks5qTnMbgaJpZM4IvTmn>
> .
>